File: MiniDOMOffice/Office/GDE/Templates/DocumentRendererBase.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica

Imports System.Drawing

Partial Class Office




    Public MustInherit Class DocumentRendererBase
        Private m_Document As CDocumento
        Private m_Template As DocumentTemplate
        Private m_Context As Object

        Public Sub New()
        End Sub

        Public Sub New(ByVal document As CDocumento, ByVal template As DocumentTemplate, ByVal context As Object)
            Me.New()
            Me.m_Document = document
            Me.m_Template = template
            Me.m_Context = context
        End Sub

        Public ReadOnly Property Document As CDocumento
            Get
                Return Me.m_Document
            End Get
        End Property

        Public ReadOnly Property DocumentTemplate As DocumentTemplate
            Get
                Return Me.m_Template
            End Get
        End Property

        Public ReadOnly Property Context As Object
            Get
                Return Me.m_Context
            End Get
        End Property

        Public MustOverride Sub Render()

        Public MustOverride Sub SaveToFile(ByVal fileName As String)


        Public Overridable Function LoadImage(ByVal path As String) As System.Drawing.Image
            path = Trim(path)
            If (path = vbNullString) Then Throw New ArgumentNullException(path)
            Return New System.Drawing.Bitmap(path)
        End Function

        Public Overridable Function GetDataField(ByVal fieldName As String) As Object
            fieldName = Trim(fieldName)
            If (fieldName = vbNullString) Then Throw New ArgumentNullException(fieldName)
            Dim t As System.Type = Me.Context.GetType
            Dim m As System.Reflection.PropertyInfo = t.GetProperty(fieldName, Reflection.BindingFlags.Public Or Reflection.BindingFlags.Instance)
            Return m.GetValue(Me.Context, New Object() {})
        End Function

        Public Overridable Function EvaluateExpression(ByVal expression As String) As Object
            Throw New NotImplementedException
        End Function


    End Class



End Class

File: MiniDOMOffice/Office/GDE/TemplateRenderer/DocumentRendererBase.cs
using System;
using System.Drawing;
using DMD;

namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Renderer dei documenti generabili
        /// </summary>
        public abstract class DocumentRendererBase
        {
            /// <summary>
            /// Documento sorgente
            /// </summary>
            private DocumentoCaricato m_Document;

            /// <summary>
            /// Template del documento
            /// </summary>
            private DocumentTemplate m_Template;

            /// <summary>
            /// Contesto in cui viene generato il documento
            /// </summary>
            private object m_Context;

            /// <summary>
            /// Costruttore
            /// </summary>
            public DocumentRendererBase()
            {
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="document">[in] Tipo di documento da generare</param>
            /// <param name="template">[in] Template da usare</param>
            /// <param name="context">[in] Contesto in cui generare il documento</param>
            public DocumentRendererBase(DocumentoCaricato document, DocumentTemplate template, object context) : this()
            {
                m_Document = document;
                m_Template = template;
                m_Context = context;
            }

            /// <summary>
            /// Tipo del documento da generare
            /// </summary>
            public DocumentoCaricato Document
            {
                get
                {
                    return m_Document;
                }
            }

            /// <summary>
            /// Template da usare per generare il documento
            /// </summary>
            public DocumentTemplate DocumentTemplate
            {
                get
                {
                    return m_Template;
                }
            }

            /// <summary>
            /// Contesto in cui generare il documento
            /// </summary>
            public object Context
            {
                get
                {
                    return m_Context;
                }
            }

            /// <summary>
            /// Genera il documento
            /// </summary>
            public abstract void Render();

            /// <summary>
            /// Genera il documento e lo salva nel file
            /// </summary>
            /// <param name="fileName"></param>
            public abstract void SaveToFile(string fileName);

            /// <summary>
            /// Carica il documento generato come un'immagine
            /// </summary>
            /// <param name="path"></param>
            /// <returns></returns>
            public virtual Image LoadImage(string path)
            {
                path = Strings.Trim(path);
                if (string.IsNullOrEmpty(path))
                    throw new ArgumentNullException(path);
                return new Bitmap(path);
            }

            /// <summary>
            /// Restituisce il valore del campo
            /// </summary>
            /// <param name="fieldName"></param>
            /// <returns></returns>
            public virtual object GetDataField(string fieldName)
            {
                //TODO migliorare GetDataField
                fieldName = Strings.Trim(fieldName);
                if (string.IsNullOrEmpty(fieldName))
                    throw new ArgumentNullException(fieldName);
                var t = Context.GetType();
                var m = t.GetProperty(fieldName, System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance);
                return m.GetValue(Context, new object[] { });
            }

            /// <summary>
            /// Valuta l'espressione
            /// </summary>
            /// <param name="expression"></param>
            /// <returns></returns>
            public virtual object EvaluateExpression(string expression)
            {
                //TODO EvaluateExpression
                throw new NotImplementedException();
            }
        }
    }
}